solidity/revert-strings-to-custom-error-migration.md

```diff
@@ -0,0 +1,39 @@
+# Purpose
+
+The purpose of this document is to standardize all error messages/handling for the Optimism Smart Contracts, which currently use a mix of error strings, and mix both `require` and `revert`.
+
+# Summary
+
+Going forward, all error messages should use the prefix of the `contract_` name and the suffix of `*Error`. Additionally, all errors should be returned via `revert` rather than `require`. This is both for visual clarity and due to the efficiency concerns around `require` in the case of errors with arguments.
+
+# Problem Statement + Context
+
+The current problem is the Optimism smart contracts mix both `require` and `revert` statements inconsistently with error strings, and Solidity custom errors. For the sake of a consistent abi, the Optimism smart contracts should converge on a single method for reverting.
+
+# Alternatives Considered
+
+## Require and Error Strings
+
+All errors could be handled using `require` statements with error strings. However, the main issue with this approach is that error strings take up significantly more bytecode size. Given the small margins of the contract size for staying within the Spurious Dragon bytecode size limit, it would be very useful to reduce contract size in favor of more optimizer runs.
+
+## Require and Custom Errors
+
+As of Solidity 0.8.24, `require` supports using custom errors, however, due to the limitation that `require` does not short-circuit (ie if there is a custom error which has a value passed in, and you pass in the result of some function call to that argument, the function call will be executed regardless of whether or not the error is returned) meaning that in all cases it is inefficient, and in some cases even dangerous.
+
+# Proposed Solution - Revert and Custom Errors
+
+All errors returned from Optimism Smart Contracts should use `revert` with custom errors for both efficiency and safety. In the case of scripts and tests which currently use `require` frequently, they should be instead replaced with `assertEq` or similar `assert*` where appropriate, including a string as a revert message. This will allow for semgrep rules to more easily enforce the no `require` constraint. Custom errors should also include the contract name as a prefix, followed by an underscore, and `Error` as a suffix for clarity and consistency.
```

Comment:
Have we verified there are no forge issues with `assertEq` when used in `setUp()` or constructors?

Reply:
I looked through open foundry issues and didn't see anything? Are there anything specific concerns you had in mind?

Reply:
This was the original issue where assertions didn't work in `setUp` and I think in constructors too: https://github.com/foundry-rs/foundry/issues/1291

That was a long time ago, but I'd suggest setting up a quick test to verify that the `assertEq` methods still cause failures when hit in both `setUp` and constructors, just to be safe

Reply:
Re-ran your POC test
```
Ran 1 test suite in 124.97ms (414.42µs CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/Counter.t.sol:SetUpFailureTest
[FAIL. Reason: setup failed: assertion failed] setUp() (gas: 0)
```

Test now fails in the setUp process which is as expect and doesn't run the other tests, LGTM?